### PR TITLE
WIP: Metrics collector for Circuit Breaker

### PR DESCRIPTION
**Pull request by slayful**

### CODE REVIEW

**Comment by whiskeysierra** [comment]
I wouldn't design/implement this with just metrics in mind. Why not add more extensive listener/callback support to the circuit breaker (similar to retries) and people can use that for metrics or anything else then.

**Comment by vgoldin** [comment]
Whenever as a generic mechanism, or specific to circuit breakers, metrics would be highly valuable. In addition, creating hystrix compatible stream would allow plugging it into hystrix dashboard

**Comment by sunng87** [comment]
I wish we can make this generic and keep failsafe self-contained. Give listener to user allows integration with any metrics provider.

**Comment by jhalterman** [comment]
@sunng87 @whiskeysierra Any particular listeners you had in mind for this?

File: src/main/java/net/jodah/failsafe/CircuitBreaker.java [hunk 1]
```diff
@@ -15,34 +15,37 @@
...
-
 import net.jodah.failsafe.function.BiPredicate;
 import net.jodah.failsafe.function.CheckedRunnable;
 import net.jodah.failsafe.function.Predicate;
-import net.jodah.failsafe.internal.CircuitBreakerStats;
-import net.jodah.failsafe.internal.CircuitState;
-import net.jodah.failsafe.internal.ClosedState;
-import net.jodah.failsafe.internal.HalfOpenState;
-import net.jodah.failsafe.internal.OpenState;
+import net.jodah.failsafe.internal.*;
```
Comment (MALPI) [inline src/main/java/net/jodah/failsafe/CircuitBreaker.java]: Probably this should be made explicit?

**Comment by MALPI** [comment]
@jhalterman I would make it more generic so that we can measure as well how long an attempt took.

File: src/main/java/net/jodah/failsafe/CircuitBreaker.java (hunk 1 above)
Comment (jhalterman) [reply src/main/java/net/jodah/failsafe/CircuitBreaker.java]: You mean the wildcard import? My IDE does this if there are several imports from the same package. The runtime cost should be nil.

**Comment by jhalterman** [comment]
@MALPI #104 currently looks to add elapsed attempt time to ExecutionContext, so we'd just need an onComplete listener for CircuitBreaker in order to access that.

This gets back to #154 where it might be nice to have separate event listeners for each Policy (such as CircuitBreaker) :)

**Comment by jhalterman** [comment]
Closing as obsolete for now.

**Pull request closed without merging**